docs/empty-statement-syntax.md

```diff
@@ -0,0 +1,86 @@
+# Empty Statement Syntax
+
+## Summary
+
+Allow for empty statements with semicolons, as introduced in Lua 5.2.
+
+```luau
+;do ; end;
+```
+Would be a syntactically valid document equivalent to:
+```luau
+do end;
+```
+
+## Motivation
+
+The semicolon symbol `;` can be used at the end of a statement, explicitly ending it. This is useful when a statement is ambiguous (especially as line breaks do not end statements), or for visual separation when several statements are in a single line. It is however not valid where the beginning of a statement would be expected, and a statement in a block can only be followed by the beginning of another statement or end of the block. Thus, a semicolon can't be used as an "empty statement" on its own. Users may expect this to be the same as an empty file, but is instead a parsing error:
+
+```luau
+;
+```
+> Error:1: Expected identifier when parsing expression, got ';'
+
+In human-written code, a deliberate empty statement is seemingly "useless", but can have intentional stylistic purposes. Mostly as intentional no-ops, visual separation of code blocks, or placeholders for future code. Lua began allowing empty statement syntax in 5.2: users of Lua may be unpleasantly surprised to find that Luau has not followed suit.
+
+In particular, who this syntax can be very useful for is tools that generate Luau code. Take this example:
+
+```luau
+print("Hello")
+GLOBAL_EVENT "Foo"
+print("World")
+```
+
+A tool may want to replace the statement `GLOBAL_EVENT "Foo"`, with for example `(_G.events["Foo"] :: Event):fire()`. Inserting it naively would cause an error due to the ambiguity of the parentheses: the first two lines are a single statement. Luau currently errors with statements like this, refusing to parse them.
+
+```luau
+print("Hello")
+(_G.events["Foo"] :: Event):fire()
+print("World")
+```
+> Error:2: Ambiguous syntax: this looks like an argument list for a function call, but could also be a start of new statement; use ';' to separate statements
+
+To solve it, the tool will likely want to insert a semicolon at the beginning, and, to be sure, end of the statement. Doing so would also be dangerous, as a syntax error would occur if the statement already had any surrounding semicolons or no statement was preceding it in the block. Tools need to check the context to skip introducing a semicolon where it'd cause an error, adding unnecessary and probably unexpected complexity. If empty statements were allowed, tools could safely surround statements with semicolons, guarding against ambiguity without having to worry about context.
```

Comment:
> the tool will likely want to insert a semicolon at the beginning, and, to be sure, end of the statement

Beginning _and_ end? Why? It's trivial to decide whether a semicolon is necessary by literally just stringifying the next statement, and then see if that string starts with `(`. If it starts with `(`, then add a `;` to the current line, then `\n` and then add the indentation and finally, the statement in question.